cmd/metal-api/internal/fsm/fsm.go

```diff
@@ -0,0 +1,304 @@
+			string(metal.ProvisioningEventPhonedHome),
+			string(metal.ProvisioningEventPlannedReboot),
+		},
+		Dst: string(metal.ProvisioningEventPhonedHome),
+	},
+	{
+		Name: string(metal.ProvisioningEventPlannedReboot),
+		Src: []string{
+			string(metal.ProvisioningEventPXEBooting),
+			string(metal.ProvisioningEventPreparing),
+			string(metal.ProvisioningEventRegistering),
+			string(metal.ProvisioningEventWaiting),
+			string(metal.ProvisioningEventInstalling),
+			string(metal.ProvisioningEventBootingNewKernel),
+			string(metal.ProvisioningEventPhonedHome),
+			string(metal.ProvisioningEventAlive),
+		},
+		Dst: string(metal.ProvisioningEventPlannedReboot),
+	},
+	{
+		Name: string(metal.ProvisioningEventAlive),
+		Src: []string{
+			string(metal.ProvisioningEventPreparing),
+		},
+		Dst: string(metal.ProvisioningEventPreparing),
+	},
+	{
+		Name: string(metal.ProvisioningEventAlive),
+		Src: []string{
+			string(metal.ProvisioningEventRegistering),
+		},
+		Dst: string(metal.ProvisioningEventRegistering),
+	},
+	{
+		Name: string(metal.ProvisioningEventAlive),
+		Src: []string{
+			string(metal.ProvisioningEventWaiting),
+		},
+		Dst: string(metal.ProvisioningEventWaiting),
+	},
+	{
+		Name: string(metal.ProvisioningEventAlive),
+		Src: []string{
+			string(metal.ProvisioningEventInstalling),
+		},
+		Dst: string(metal.ProvisioningEventInstalling),
+	},
+	{
+		Name: string(metal.ProvisioningEventAlive),
+		Src: []string{
+			string(metal.ProvisioningEventBootingNewKernel),
+		},
+		Dst: string(metal.ProvisioningEventBootingNewKernel),
+	},
+}
+
+// every callback expects parameters of types *EventContainer, *metal.ProvisioningEvent, *zap.SugaredLogger
+var callbacks = fsm.Callbacks{
+	"enter_state": handleStateTransition,
+	"enter_" + string(metal.ProvisioningEventPlannedReboot): handlePlannedRebootEvent,
+	"enter_" + string(metal.ProvisioningEventPhonedHome):    handlePhonedHomeEvent,
+	"enter_" + string(metal.ProvisioningEventAlive):         handleAliveEvent,
+}
+
+func handleStateTransition(e *fsm.Event) {
+	container, event, _, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	if e.Event != string(metal.ProvisioningEventPhonedHome) {
+		container.Events = append(container.Events, *event)
+		container.LastEventTime = &(*event).Time
+		container.Liveliness = metal.MachineLivelinessAlive
+	}
+}
+
+func handlePlannedRebootEvent(e *fsm.Event) {
+	container, _, _, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	container.IncompleteCycles = 0
+	container.Liveliness = metal.MachineLivelinessAlive
+}
+
+func handlePhonedHomeEvent(e *fsm.Event) {
+	container, event, _, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	if e.Src == string(metal.ProvisioningEventPlannedReboot) {
+		if event.Time.Sub(*container.LastEventTime) > timeOutAfterPlannedReboot {
+			container.Liveliness = metal.MachineLivelinessUnknown
+		}
```

Comment:
Is a bit different: The problematic state only comes up when the machine is being freed (reclaimed) and then boots back up into the user's OS emitting Phoned Home events again. Instead maybe introduce a new state `MachineReclaim` which is emitted during machine free.